demo/ota-agent/main.c

```diff
@@ -0,0 +1,236 @@
+
+static void mqttEventCallback( MQTTContext_t * mqttContext,
+                               MQTTPacketInfo_t * packetInfo,
+                               MQTTDeserializedInfo_t * deserializedInfo );
+
+static void handleIncomingMQTTMessage( char * topic,
+                                       size_t topicLength,
+                                       uint8_t * message,
+                                       size_t messageLength );
+
+int main( int argc, char * argv[] )
+{
+    MQTTStatus_t mqttResult;
+    MQTTFixedBuffer_t fixedBuffer = { 0 };
+
+    if( argc != 6 )
+    {
+        printf( "Usage: %s certificateFilePath privateKeyFilePath "
+                "rootCAFilePath endpoint thingName\n",
+                argv[ 0 ] );
+        return 1;
+    }
+
+    MQTTAgentLock = xSemaphoreCreateRecursiveMutexStatic(
+        &MQTTAgentLockBuffer );
+    MQTTStateUpdateLock = xSemaphoreCreateMutexStatic(
+        &MQTTStateUpdateLockBuffer );
+
+    fixedBuffer.pBuffer = networkBuffer;
+    fixedBuffer.size = 5000U;
+
+    transport_tlsInit( &transport );
+    mqttResult = MQTT_Init( &mqttContext,
+                            &transport,
+                            Clock_GetTimeMs,
+                            mqttEventCallback,
+                            &fixedBuffer );
+    assert( mqttResult == MQTTSuccess );
+
+    xTaskCreate( otaAgentTask, "T_OTA", 6000, ( void * ) argv, 1, NULL );
+    xTaskCreate( mqttProcessLoopTask, "T_MQTT", 6000, NULL, 2, NULL );
+    xTaskCreate( suspendResumeLoopTask, "T_SUSPEND", 6000, NULL, 2, NULL );
+
+    mqttWrapper_setCoreMqttContext( &mqttContext );
+    mqttWrapper_setThingName( argv[ 5 ],
+                              strnlen( argv[ 5 ], MAX_THING_NAME_SIZE ) );
+
+    vTaskStartScheduler();
+
+    return 0;
+}
+
+static void suspendResumeLoopTask( void * parameters )
+{
+    ( void ) parameters;
+    OtaEventMsg_t testEvent = { 0 };
+    OtaState_t curState;
+    printf("Suspend resume task \n");
+    while( true )
+    {
+        curState = getOtaAgentState();
+
+        if  (curState != OtaAgentStateStopped && curState >= OtaAgentStateRequestingJob && curState != OtaAgentStateSuspended)
+        {
+            testEvent.eventId = OtaAgentEventSuspend;
+            OtaSendEvent_FreeRTOS( &testEvent );
+        }
+
+        vTaskDelay( 10 );
+
+        curState = getOtaAgentState();
+        if (curState == OtaAgentStateSuspended)
+        {
+            testEvent.eventId = OtaAgentEventResume;
+            OtaSendEvent_FreeRTOS( &testEvent );
+        }
+        vTaskDelay( 30 );
+
+    }
+}
+
+static void mqttProcessLoopTask( void * parameters )
+{
+    ( void ) parameters;
+
+    while( true )
+    {
+        if( mqttWrapper_isConnected() )
+        {
+            MQTTStatus_t status = MQTT_ProcessLoop( &mqttContext );
+
+            if( status == MQTTRecvFailed )
+            {
+                printf( "ERROR: MQTT Receive failed. Closing connection.\n" );
+                exit( 1 );
```

Comment:
Is this closing the connection? Seems like it would kill the demo

Reply:
Yes, this can kill the demo. We followed the same approach in the chain of responsibility demo as well.
We can add retries if don't want to stop the demo in case of network failure or connection issues.